### PR TITLE
FIX: Animation desyncs during multi-step actions and attack prompts

### PR DESCRIPTION
**Pull request by Gloomy26**
## Description

- Animations could get stuck behind prompts during multi-step actions, causing visual desyncs. Instead of scattering manual flushes across individual card scripts, animation syncing is now centralized at the engine level.

- Added `flush_choreography` to the targeted prompt wrappers that handle multi-step actions: `choose_cards`, `reveal_hand`, `discard_energy_units_from`, `place_damage_counters`, and `choose_attack_to_copy`. I tested all these one by one and left out the ones that don't need flushing.

- Added an `attack_bracket_sent` flag to `resolve_attack` and `flush_choreography`, allowing animations to be flushed early for mid-attack prompts without sending the attack bracket twice at the end.

- Removed `flush_choreography` from card scripts where it was no longer necessary. More complex cards will still need manual flushing, but that's to be expected. Cram-o-matic is a good example; I had to add a flush to it so that the coinflip would trigger only after a card had been discarded.

## Footage (required)
While this fix affects many interactions, I've recorded just two examples showing the before/after behavior. I can provide more if needed.

<details>
<summary><b>Before</b></summary>

https://github.com/user-attachments/assets/d580851a-0409-48de-b306-67300cc87e0f

https://github.com/user-attachments/assets/48c94b3e-3eb4-4623-a494-d45d44228f88

</details>

<details>
<summary><b>After</b></summary>

https://github.com/user-attachments/assets/e8d14742-54f9-48bc-8d84-a5d2085694c0

https://github.com/user-attachments/assets/3bdb9661-8bb9-4305-b937-23321a1de7c5

</details>

## Submission checklist

- [x] I have described my changes and what they do.
- [x] I have included footage demonstrating my changes.

### CODE REVIEW

File: spirit/game/session/effects.py [hunk 1]
```diff
@@ -2304,6 +2324,10 @@ async def resolve_attack(session, player_id: str, attacker: PokemonEntity,
     ctx = AttackContext(session, player_id, attacker, ability)
     effect = ability.effect if ability else None
     title = ability.title if ability else action_id
+    # Store these on context in case a mid-attack prompt needs to flush early.
+    ctx.action_id = action_id
+    ctx.title = title
+    ctx.attack_bracket_sent = False
```
Comment (Bratah123) [inline spirit/game/session/effects.py:2333]:
```py
# Store these on context in case a mid-attack prompt needs to flush early.
ctx.action_id = action_id
ctx.title = title
ctx.attack_bracket_sent = False
```

These should be defined in

```py
class EffectContext:
```
> Add comments on how they're used

File: spirit/game/session/effects.py [hunk 2]
```diff
@@ -2070,10 +2082,18 @@ async def switch_active(self, player_id: str, new_active: PokemonEntity) -> bool
         return True
 
     async def flush_choreography(self):
+        if not self._messages:
+            return
+        # If flushing mid-attack, fire the initial attack bracket first so 
+        # animations aren't trapped behind prompts.
+        if getattr(self, "attack_bracket_sent", True) is False:
```
Comment (Bratah123) [inline spirit/game/session/effects.py]: you can just do:

```py
if not self.attack_bracket_sent:
```

After it has been defined in ``EffectContext`` (see other comment)

**Review by Bratah123 — CHANGES_REQUESTED** [review]

File: spirit/game/session/effects.py [hunk 3]
```diff
@@ -2324,10 +2328,6 @@ async def resolve_attack(session, player_id: str, attacker: PokemonEntity,
     ctx = AttackContext(session, player_id, attacker, ability)
     effect = ability.effect if ability else None
     title = ability.title if ability else action_id
-    # Store these on context in case a mid-attack prompt needs to flush early.
-    ctx.action_id = action_id
-    ctx.title = title
```
Comment (Bratah123) [inline spirit/game/session/effects.py]: Should these assignments stay? where else is action_id and title assigned

Comment (Gloomy26) [reply spirit/game/session/effects.py]: Oh I'm sorry I meant to remove the comment only I don't know why I removed the whole thing. I'll fix it rn

**Comment by Gloomy26** [comment]
After addressing the bug caused by moving definitions to EffectContext, I ran a few more tests and that helped me realize that flushing early in some prompt wrappers can introduce edge cases with complex, multi-step cards (with 2, 3 or more nested prompts). While flushing early benefits a lot of standard cards, it's probably safer to hold off on that global change for now to avoid unintended side effects. I'm going to close this PR and open a new one that addresses just the bug in send_attack_bracket. Does that sound good? I apologize for creating such back and forth!
